README.md
# NAME

Module::CPANfile - Parse cpanfile

# SYNOPSIS

    use Module::CPANfile;

    my $file = Module::CPANfile->load("cpanfile");
    my $prereqs = $file->prereqs; # CPAN::Meta::Prereqs object

    my @features = $file->features; # CPAN::Meta::Feature objects
    my $merged_prereqs = $file->prereqs_with(@identifiers); # CPAN::Meta::Prereqs

    $file->merge_meta('MYMETA.json');

# DESCRIPTION

Module::CPANfile is a tool to handle [cpanfile](https://metacpan.org/pod/cpanfile) format to load application
specific dependencies, not just for CPAN distributions.

# METHODS

- load

        $file = Module::CPANfile->load;
        $file = Module::CPANfile->load('cpanfile');

    Load and parse a cpanfile. By default it tries to load `cpanfile` in
    the current directory, unless you pass the path to its argument.

- from\_prereqs

        $file = Module::CPANfile->from_prereqs({
          runtime => { requires => { DBI => '1.000' } },
        });

    Creates a new Module::CPANfile object from prereqs hash you can get
    via [CPAN::Meta](https://metacpan.org/pod/CPAN::Meta)'s `prereqs`, or [CPAN::Meta::Prereqs](https://metacpan.org/pod/CPAN::Meta::Prereqs)'
    `as_string_hash`.

        # read MYMETA, then feed the prereqs to create Module::CPANfile
        my $meta = CPAN::Meta->load_file('MYMETA.json');
        my $file = Module::CPANfile->from_prereqs($meta->prereqs);

        # load cpanfile, then recreate it with round-trip
        my $file = Module::CPANfile->load('cpanfile');
        $file = Module::CPANfile->from_prereqs($file->prereq_specs);
                                          # or $file->prereqs->as_string_hash

- prereqs

    Returns [CPAN::Meta::Prereqs](https://metacpan.org/pod/CPAN::Meta::Prereqs) object out of the parsed cpanfile.

- prereq\_specs

    Returns a hash reference that should be passed to `CPAN::Meta::Prereqs->new`.

- features

    Returns a list of features available in the cpanfile as [CPAN::Meta::Feature](https://metacpan.org/pod/CPAN::Meta::Feature).

- prereqs\_with(@identifiers), effective\_prereqs(\\@identifiers)

    Returns [CPAN::Meta::Prereqs](https://metacpan.org/pod/CPAN::Meta::Prereqs) object, with merged prereqs for
    features identified with the `@identifiers`.

- to\_string($include\_empty)

        $file->to_string;
        $file->to_string(1);

    Returns a canonical string (code) representation for cpanfile. Useful
    if you want to convert [CPAN::Meta::Prereqs](https://metacpan.org/pod/CPAN::Meta::Prereqs) to a new cpanfile.

        # read MYMETA's prereqs and print cpanfile representation of it
        my $meta = CPAN::Meta->load_file('MYMETA.json');
        my $file = Module::CPANfile->from_prereqs($meta->prereqs);
        print $file->to_string;

    By default, it omits the phase where there're no modules
    registered. If you pass the argument of a true value, it will print
    them as well.

- save

        $file->save('cpanfile');

    Saves the currently loaded prereqs as a new `cpanfile` by calling
    `to_string`. Beware **this method will overwrite the existing
    cpanfile without any warning or backup**. Taking a backup or giving
    warnings to users is a caller's responsibility.

        # Read MYMETA.json and creates a new cpanfile
        my $meta = CPAN::Meta->load_file('MYMETA.json');
        my $file = Module::CPANfile->from_prereqs($meta->prereqs);
        $file->save('cpanfile');

- merge\_meta

        $file->merge_meta('META.yml');
        $file->merge_meta('MYMETA.json', '2.0');

    Merge the effective prereqs with Meta specification loaded from the
    given META file, using CPAN::Meta. You can specify the META spec
    version in the second argument, which defaults to 1.4 in case the
    given file is YAML, and 2 if it is JSON.

- options\_for\_module

        my $options = $file->options_for_module($module);

    Returns the extra options specified for a given module as a hash
    reference. Returns `undef` when the given module is not specified in
    the `cpanfile`.

    For example,

        # cpanfile
        requires 'Plack', '1.000',
          dist => "MIYAGAWA/Plack-1.000.tar.gz";

        # ...
        my $file = Module::CPANfile->load;
        my $options = $file->options_for_module('Plack');
        # => { dist => "MIYAGAWA/Plack-1.000.tar.gz" }

# AUTHOR

Tatsuhiko Miyagawa

# SEE ALSO

[cpanfile](https://metacpan.org/pod/cpanfile), [CPAN::Meta](https://metacpan.org/pod/CPAN::Meta), [CPAN::Meta::Spec](https://metacpan.org/pod/CPAN::Meta::Spec)


Changes
Revision history for Module::Install::CPANfile

{{$NEXT}}

1.1004  2018-04-26 02:07:14 PDT
        - Fix a bug in to_strings with empty feature block (charsbar) #50
        - Explicitly add . in tests for perl 5.26

1.1003  2018-04-22 01:54:46 CEST
        - Make options written back with save()
        - Documented options_for_module method
        - Internal refactoring

1.1002  2016-02-12 08:23:00 PST
        - Make it work under the taint mode (-T) #41

1.1001  2015-09-04 14:07:14 CEST
        - Fixed tests to avoid using subtests. #38

1.1000  2014-08-29 13:48:13 JST
        - Make it non-development release

1.0905  2013-12-09 16:26:08 PST
        - Support mirror syntax (masaki) #22

1.0904  2013-09-21 07:56:31 JST
        - Removed git/ref accessor from Requirement
        - Added $requirement->has_options

1.0903  2013-09-20 15:36:30 JST
        - Added $cpanfile->merged_requirements to get the CPAN::Meta::Requirements for all prereqs

1.0902  2013-09-20 13:45:15 JST
        - s/rev/ref/ for git options

1.0901  2013-09-19 18:59:55 JST
        - Remove unused code

1.0900  2013-09-19 18:45:59 JST
        - Experimental support for git URL specification
        - Complete rewrite of internal to preserve original prereq statement as much as possible

1.0002  2013-09-06 12:26:11 PDT
        - Add cpanfile-dump utility (xdg)

1.0001  2013-08-05 14:24:07 PDT
        - Updated documentation about syntax

1.0000  2013-08-04 12:28:32 PDT
        - Make the cpanfile spec 1.0
        - Updated documentation

0.9036  2013-07-24 13:11:42 PDT
        - Repackage after git rebase

0.9035  2013-07-24 13:10:29 PDT
        - Fix a bug where warnings will be raised if the content of cpanfile has sprintf-style
          %X content somewhere (reported by frew)

0.9034  2013-06-05 12:09:11 JST
        - split .pm files, mainly for PAUSE

0.9033  2013-06-05 11:38:56 JST
        - merge with git master

0.9032  2013-06-05 11:35:58 JST
        - repackage for PAUSE

0.9031  2013-04-17 08:39:17 JST
        - Revert EXPERIMENTAL git URL options since it causes failures when stringified
          inside CPAN::Meta #13

0.9030  2013-04-14 17:05:46 JST
        - Fixed a bug where prereqs_with with an empty feature list causes an exception

0.9029  2013-04-14 13:05:30 JST
        - Support EXPERIMENTAL git URL parsing (ikasam_a)
        - Allow feature DSL to omit description
        - Add better error messages for feature syntax errors

0.9028  2013-04-14 01:27:29 JST
        - Add new 'feature' DSL and features support for CPAN::Meta::Feature

0.9027  2013-03-31 12:45:10 PDT
        - Improved error message shows filename and line number

0.9026  2013-03-28 12:07:27 PDT
        - Updated FAQ to mention tools that support cpanfile

0.9025  2013-03-24 23:27:57 PDT
        - rebuild package with new Milla to support script

0.9024  2013-03-23 21:29:41 PDT
        - Make it non-trial

0.9023  2013-03-23 21:26:07 PDT
        - mymeta-cpanfile: Added options to filter specific phases and types

0.9022  2013-03-23 20:34:30 America/Los_Angeles
        - Convert to Milla
        - Made CPAN::Meta and ::Prereqs a hard requirement than runtime optional

0.9021  Sat Mar 23 09:11:46 PDT 2013
        - added mymeta-cpanfile utility script for migration from Makefile.PL/Build.PL

0.9020  Sat Mar 23 02:48:50 PDT 2013
        - Implemeneted from_prereqs, to_string and save utility methods,
          to allow creating a new cpanfile out of (MY)META files

0.9010  Fri Feb 15 13:24:40 PST 2013
        - Fix the package name in the example

0.9009  Thu Feb  7 09:24:11 PST 2013
        - Fixed CPAN::Meta dependency in the test (Chris Weyl)

0.9008  Wed Jan 30 21:53:32 PST 2013
        - Added merge_meta method to be used in Module::Install::CPANfile

0.9007  Sat Apr 14 14:55:49 CST 2012
        - Fixed documentation

0.9006  Fri Apr 13 07:30:33 JST 2012
        - Alias prereqs method to prereq

0.9005  Thu Apr 12 20:41:27 JST 2012
        - Include all files into Module/CPANfile.pm

0.9004  Thu Apr 12 18:31:59 JST 2012
        - Renamed the distribution

0.9003  Sat Apr  7 01:26:04 PDT 2012
        - Don't use $! in error checks because we don't open file there.
          This also caused issues in 5.8 where $! isn't often cleared

0.9002  Sat Apr  7 01:14:00 PDT 2012
        - Fixed the test again to not fail on 5.8

0.9001  Tue Apr  3 04:30:31 CEST 2012
        - Added t/samples in the directory so the test won't fail

0.9000  Sun Apr  1 15:10:36 CEST 2012
        - Initial version


lib/Module/CPANfile.pm
package Module::CPANfile;
use strict;
use warnings;
use Cwd;
use Carp ();
use Module::CPANfile::Environment;
use Module::CPANfile::Requirement;

our $VERSION = '1.1003';

BEGIN {
    if (${^TAINT}) {
        *untaint = sub {
            my $str = shift;
            ($str) = $str =~ /^(.+)$/s;
            $str;
        };
    } else {
        *untaint = sub { $_[0] };
    }
}

sub new {
    my($class, $file) = @_;
    bless {}, $class;
}

sub load {
    my($proto, $file) = @_;

    my $self = ref $proto ? $proto : $proto->new;
    $self->parse($file || _default_cpanfile());
    $self;
}

sub save {
    my($self, $path) = @_;

    open my $out, ">", $path or die "$path: $!";
    print {$out} $self->to_string;
}

sub parse {
    my($self, $file) = @_;

    my $code = do {
        open my $fh, "<", $file or die "$file: $!";
        join '', <$fh>;
    };

    $code = untaint $code;

    my $env = Module::CPANfile::Environment->new($file);
    $env->parse($code) or die $@;

    $self->{_mirrors} = $env->mirrors;
    $self->{_prereqs} = $env->prereqs;
}

sub from_prereqs {
    my($proto, $prereqs) = @_;

    my $self = $proto->new;
    $self->{_prereqs} = Module::CPANfile::Prereqs->from_cpan_meta($prereqs);

    $self;
}

sub mirrors {
    my $self = shift;
    $self->{_mirrors} || [];
}

sub features {
    my $self = shift;
    map $self->feature($_), $self->{_prereqs}->identifiers;
}

sub feature {
    my($self, $identifier) = @_;
    $self->{_prereqs}->feature($identifier);
}

sub prereq { shift->prereqs }

sub prereqs {
    my $self = shift;
    $self->{_prereqs}->as_cpan_meta;
}

sub merged_requirements {
    my $self = shift;
    $self->{_prereqs}->merged_requirements;
}

sub effective_prereqs {
    my($self, $features) = @_;
    $self->prereqs_with(@{$features || []});
}

sub prereqs_with {
    my($self, @feature_identifiers) = @_;

    my @others = map { $self->feature($_)->prereqs } @feature_identifiers;
    $self->prereqs->with_merged_prereqs(\@others);
}

sub prereq_specs {
    my $self = shift;
    $self->prereqs->as_string_hash;
}

sub prereq_for_module {
    my($self, $module) = @_;
    $self->{_prereqs}->find($module);
}

sub options_for_module {
    my($self, $module) = @_;
    my $prereq = $self->prereq_for_module($module) or return;
    $prereq->requirement->options;
}

sub merge_meta {
    my($self, $file, $version) = @_;

    require CPAN::Meta;

    $version ||= $file =~ /\.yml$/ ? '1.4' : '2';

    my $prereq = $self->prereqs;

    my $meta = CPAN::Meta->load_file($file);
    my $prereqs_hash = $prereq->with_merged_prereqs($meta->effective_prereqs)->as_string_hash;
    my $struct = { %{$meta->as_struct}, prereqs => $prereqs_hash };

    CPAN::Meta->new($struct)->save($file, { version => $version });
}

sub _d($) {
    require Data::Dumper;
    chomp(my $value = Data::Dumper->new([$_[0]])->Terse(1)->Dump);
    $value;
}

sub _default_cpanfile {
    my $file = Cwd::abs_path('cpanfile');
    untaint $file;
}

sub to_string {
    my($self, $include_empty) = @_;

    my $mirrors = $self->mirrors;
    my $prereqs = $self->prereq_specs;

    my $code = '';
    $code .= $self->_dump_mirrors($mirrors);
    $code .= $self->_dump_prereqs($prereqs, $include_empty);

    for my $feature ($self->features) {
        $code .= "feature @{[ _d $feature->{identifier} ]}, @{[ _d $feature->{description} ]} => sub {\n";
        $code .= $self->_dump_prereqs($feature->{prereqs}->as_string_hash, $include_empty, 4);
        $code .= "};\n\n";
    }

    $code =~ s/\n+$/\n/s;
    $code;
}

sub _dump_mirrors {
    my($self, $mirrors) = @_;

    my $code = "";

    for my $url (@$mirrors) {
        $code .= "mirror @{[ _d $url ]};\n";
    }

    $code =~ s/\n+$/\n/s;
    $code;
}

sub _dump_prereqs {
    my($self, $prereqs, $include_empty, $base_indent) = @_;

    my $code = '';
    for my $phase (qw(runtime configure build test develop)) {
        my $indent = $phase eq 'runtime' ? '' : '    ';
        $indent .= (' ' x ($base_indent || 0));

        my($phase_code, $requirements);
        $phase_code .= "on $phase => sub {\n" unless $phase eq 'runtime';

        for my $type (qw(requires recommends suggests conflicts)) {
            for my $mod (sort keys %{$prereqs->{$phase}{$type}}) {
                my $ver = $prereqs->{$phase}{$type}{$mod};
                $phase_code .= $ver eq '0'
                             ? "${indent}$type @{[ _d $mod ]}"
                             : "${indent}$type @{[ _d $mod ]}, @{[ _d $ver ]}";

                my $options = $self->options_for_module($mod) || {};
                if (%$options) {
                    my @opts;
                    for my $key (keys %$options) {
                        my $k = $key =~ /^[a-zA-Z0-9_]+$/ ? $key : _d $key;
                        push @opts, "$k => @{[ _d $options->{$k} ]}";
                    }

                    $phase_code .= ",\n" . join(",\n", map "  $indent$_", @opts);
                }

                $phase_code .= ";\n";
                $requirements++;
            }
        }

        $phase_code .= "\n" unless $requirements;
        $phase_code .= "};\n" unless $phase eq 'runtime';

        $code .= $phase_code . "\n" if $requirements or $include_empty;
    }

    $code =~ s/\n+$/\n/s;
    $code;
}

1;

__END__

=head1 NAME

Module::CPANfile - Parse cpanfile

=head1 SYNOPSIS

  use Module::CPANfile;

  my $file = Module::CPANfile->load("cpanfile");
  my $prereqs = $file->prereqs; # CPAN::Meta::Prereqs object

  my @features = $file->features; # CPAN::Meta::Feature objects
  my $merged_prereqs = $file->prereqs_with(@identifiers); # CPAN::Meta::Prereqs

  $file->merge_meta('MYMETA.json');

=head1 DESCRIPTION

Module::CPANfile is a tool to handle L<cpanfile> format to load application
specific dependencies, not just for CPAN distributions.

=head1 METHODS

=over 4

=item load

  $file = Module::CPANfile->load;
  $file = Module::CPANfile->load('cpanfile');

Load and parse a cpanfile. By default it tries to load C<cpanfile> in
the current directory, unless you pass the path to its argument.

=item from_prereqs

  $file = Module::CPANfile->from_prereqs({
    runtime => { requires => { DBI => '1.000' } },
  });

Creates a new Module::CPANfile object from prereqs hash you can get
via L<CPAN::Meta>'s C<prereqs>, or L<CPAN::Meta::Prereqs>'
C<as_string_hash>.

  # read MYMETA, then feed the prereqs to create Module::CPANfile
  my $meta = CPAN::Meta->load_file('MYMETA.json');
  my $file = Module::CPANfile->from_prereqs($meta->prereqs);

  # load cpanfile, then recreate it with round-trip
  my $file = Module::CPANfile->load('cpanfile');
  $file = Module::CPANfile->from_prereqs($file->prereq_specs);
                                    # or $file->prereqs->as_string_hash

=item prereqs

Returns L<CPAN::Meta::Prereqs> object out of the parsed cpanfile.

=item prereq_specs

Returns a hash reference that should be passed to C<< CPAN::Meta::Prereqs->new >>.

=item features

Returns a list of features available in the cpanfile as L<CPAN::Meta::Feature>.

=item prereqs_with(@identifiers), effective_prereqs(\@identifiers)

Returns L<CPAN::Meta::Prereqs> object, with merged prereqs for
features identified with the C<@identifiers>.

=item to_string($include_empty)

  $file->to_string;
  $file->to_string(1);

Returns a canonical string (code) representation for cpanfile. Useful
if you want to convert L<CPAN::Meta::Prereqs> to a new cpanfile.

  # read MYMETA's prereqs and print cpanfile representation of it
  my $meta = CPAN::Meta->load_file('MYMETA.json');
  my $file = Module::CPANfile->from_prereqs($meta->prereqs);
  print $file->to_string;

By default, it omits the phase where there're no modules
registered. If you pass the argument of a true value, it will print
them as well.

=item save

  $file->save('cpanfile');

Saves the currently loaded prereqs as a new C<cpanfile> by calling
C<to_string>. Beware B<this method will overwrite the existing
cpanfile without any warning or backup>. Taking a backup or giving
warnings to users is a caller's responsibility.

  # Read MYMETA.json and creates a new cpanfile
  my $meta = CPAN::Meta->load_file('MYMETA.json');
  my $file = Module::CPANfile->from_prereqs($meta->prereqs);
  $file->save('cpanfile');

=item merge_meta

  $file->merge_meta('META.yml');
  $file->merge_meta('MYMETA.json', '2.0');

Merge the effective prereqs with Meta specification loaded from the
given META file, using CPAN::Meta. You can specify the META spec
version in the second argument, which defaults to 1.4 in case the
given file is YAML, and 2 if it is JSON.

=item options_for_module

  my $options = $file->options_for_module($module);

Returns the extra options specified for a given module as a hash
reference. Returns C<undef> when the given module is not specified in
the C<cpanfile>.

For example,

  # cpanfile
  requires 'Plack', '1.000',
    dist => "MIYAGAWA/Plack-1.000.tar.gz";

  # ...
  my $file = Module::CPANfile->load;
  my $options = $file->options_for_module('Plack');
  # => { dist => "MIYAGAWA/Plack-1.000.tar.gz" }

=back

=head1 AUTHOR

Tatsuhiko Miyagawa

=head1 SEE ALSO

L<cpanfile>, L<CPAN::Meta>, L<CPAN::Meta::Spec>

=cut
